File: cronista/client/api_client.py
"""Cliente HTTP para a API. Ver docs/09-api.md."""

from __future__ import annotations

import httpx2

from cronista.client import token_store
from cronista.client.settings import ClientSettings

settings = ClientSettings()


class ApiError(Exception):
    def __init__(self, message: str, status_code: int | None = None) -> None:
        super().__init__(message)
        self.status_code = status_code


def _post(
    path: str, json_body: dict, access_token: str | None = None, timeout: float = 10.0
) -> httpx2.Response:
    headers = {"Authorization": f"Bearer {access_token}"} if access_token else {}
    try:
        return httpx2.post(
            f"{settings.api_base_url}{path}", json=json_body, headers=headers, timeout=timeout
        )
    except httpx2.ConnectError as exc:
        raise ApiError(
            "Não foi possível conectar à API. Verifique se o container está no ar."
        ) from exc


def _get(
    path: str, params: dict | None = None, access_token: str | None = None, timeout: float = 10.0
) -> httpx2.Response:
    headers = {"Authorization": f"Bearer {access_token}"} if access_token else {}
    try:
        return httpx2.get(
            f"{settings.api_base_url}{path}", params=params, headers=headers, timeout=timeout
        )
    except httpx2.ConnectError as exc:
        raise ApiError(
            "Não foi possível conectar à API. Verifique se o container está no ar."
        ) from exc


def _patch(path: str, json_body: dict, access_token: str | None = None, timeout: float = 10.0) -> httpx2.Response:
    headers = {"Authorization": f"Bearer {access_token}"} if access_token else {}
    try:
        return httpx2.patch(
            f"{settings.api_base_url}{path}", json=json_body, headers=headers, timeout=timeout
        )
    except httpx2.ConnectError as exc:
        raise ApiError(
            "Não foi possível conectar à API. Verifique se o container está no ar."
        ) from exc


def _delete(path: str, access_token: str | None = None, timeout: float = 10.0) -> httpx2.Response:
    headers = {"Authorization": f"Bearer {access_token}"} if access_token else {}
    try:
        return httpx2.delete(f"{settings.api_base_url}{path}", headers=headers, timeout=timeout)
    except httpx2.ConnectError as exc:
        raise ApiError(
            "Não foi possível conectar à API. Verifique se o container está no ar."
        ) from exc


def _raise_for_status(resp: httpx2.Response) -> None:
    try:
        resp.raise_for_status()
    except httpx2.HTTPStatusError as exc:
        raise ApiError(str(exc), status_code=resp.status_code) from exc


def login(username: str, password: str) -> dict[str, str]:
    resp = _post("/auth/login", {"username": username, "password": password})
    if resp.status_code == 401:
        raise ApiError("Usuário ou senha incorretos.", status_code=401)
    _raise_for_status(resp)
    return resp.json()


def refresh(refresh_token: str) -> str:
    resp = _post("/auth/refresh", {"refresh_token": refresh_token})
    if resp.status_code == 401:
        raise ApiError("Sessão expirada. Faça login novamente.", status_code=401)
    _raise_for_status(resp)
    return resp.json()["access_token"]


def _authed_post(path: str, json_body: dict, timeout: float = 10.0) -> dict:
    """POST autenticado com renovação silenciosa (docs/11-cli.md §3): um
    401 tenta `refresh` uma vez antes de desistir, sem pedir senha de novo."""
    tokens = token_store.load_tokens()
    if tokens is None:
        raise ApiError("Não autenticado. Rode `cronista login`.", status_code=401)

    resp = _post(path, json_body, tokens["access_token"], timeout=timeout)
    if resp.status_code == 401:
        new_access_token = refresh(tokens["refresh_token"])
        token_store.save_tokens(new_access_token, tokens["refresh_token"])
        resp = _post(path, json_body, new_access_token, timeout=timeout)

    _raise_for_status(resp)
    return resp.json()


def _authed_get(path: str, params: dict | None = None, timeout: float = 10.0) -> dict:
    """GET autenticado com renovação silenciosa -- mesmo padrão de
    `_authed_post` (docs/11-cli.md §3)."""
    tokens = token_store.load_tokens()
    if tokens is None:
        raise ApiError("Não autenticado. Rode `cronista login`.", status_code=401)

    resp = _get(path, params, tokens["access_token"], timeout=timeout)
    if resp.status_code == 401:
        new_access_token = refresh(tokens["refresh_token"])
        token_store.save_tokens(new_access_token, tokens["refresh_token"])
        resp = _get(path, params, new_access_token, timeout=timeout)

    _raise_for_status(resp)
    return resp.json()


def _authed_patch(path: str, json_body: dict, timeout: float = 10.0) -> dict:
    """PATCH autenticado com renovação silenciosa -- mesmo padrão de
    `_authed_post` (docs/11-cli.md §3)."""
    tokens = token_store.load_tokens()
    if tokens is None:
        raise ApiError("Não autenticado. Rode `cronista login`.", status_code=401)

    resp = _patch(path, json_body, tokens["access_token"], timeout=timeout)
    if resp.status_code == 401:
        new_access_token = refresh(tokens["refresh_token"])
        token_store.save_tokens(new_access_token, tokens["refresh_token"])
        resp = _patch(path, json_body, new_access_token, timeout=timeout)

    _raise_for_status(resp)
    return resp.json()


def _authed_delete(path: str, timeout: float = 10.0) -> None:
    """DELETE autenticado com renovação silenciosa -- mesmo padrão de
    `_authed_patch` (docs/11-cli.md §3). Sem `.json()` no retorno: 204
    não tem corpo."""
    tokens = token_store.load_tokens()
    if tokens is None:
        raise ApiError("Não autenticado. Rode `cronista login`.", status_code=401)

    resp = _delete(path, tokens["access_token"], timeout=timeout)
    if resp.status_code == 401:
        new_access_token = refresh(tokens["refresh_token"])
        token_store.save_tokens(new_access_token, tokens["refresh_token"])
        resp = _delete(path, new_access_token, timeout=timeout)

    _raise_for_status(resp)


def create_meeting(payload: dict) -> dict:
    return _authed_post("/meetings", payload)


def register_track(meeting_id: object, payload: dict) -> dict:
    return _authed_post(f"/meetings/{meeting_id}/tracks", payload)


def reprocessar(meeting_id: object) -> dict:
    """UC-05, RF-15 (docs/12-transcricao.md §10)."""
    return _authed_post(f"/meetings/{meeting_id}/transcribe", {})


def resumir(meeting_id: object) -> dict:
    """UC-06, RF-16. Bloqueia até o provedor responder -- timeout bem
    maior que o padrão, um LLM local demora mais que 10s pra gerar."""
    return _authed_post(f"/meetings/{meeting_id}/summarize", {}, timeout=300.0)


def list_meetings() -> list[dict]:
    """RF-20, UC-07."""
    return _authed_get("/meetings")


def get_meeting(meeting_id: object) -> dict:
    """RF-20/RF-22, UC-07: dados da reunião com trilhas e resumos."""
    return _authed_get(f"/meetings/{meeting_id}")


def get_transcript(meeting_id: object) -> list[dict]:
    """RF-21, UC-07."""
    return _authed_get(f"/meetings/{meeting_id}/transcript")


def rename_meeting(meeting_id: object, title: str) -> dict:
    """UC-07."""
    return _authed_patch(f"/meetings/{meeting_id}", {"title": title})


def delete_meeting(meeting_id: object) -> None:
    """RF-30, UC-09. A confirmação é responsabilidade de quem chama
    (CLI) -- esta função, uma vez chamada, remove sem mais perguntas."""
    _authed_delete(f"/meetings/{meeting_id}")


def search(
    q: str, speaker: str | None = None, since: str | None = None, until: str | None = None
) -> list[dict]:
    """RF-23/RF-24, UC-08."""
    params = {"q": q}
    if speaker is not None:
        params["speaker"] = speaker
    if since is not None:
        params["since"] = since
    if until is not None:
        params["until"] = until
    return _authed_get("/search", params)

File: cronista-tui/src/apiClient.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.join(__dirname, "..", "..");

export class ApiError extends Error {
  constructor(message, statusCode) {
    super(message);
    this.statusCode = statusCode;
  }
}

// Lida uma vez só: API_BASE_URL não muda no meio do processo, então não
// há por que reabrir e reler o .env do disco a cada chamada de API.
let _apiBaseUrl = null;

function readApiBaseUrl() {
  if (_apiBaseUrl !== null) return _apiBaseUrl;
  // Variável de ambiente tem prioridade sobre o .env, igual
  // cronista/client/settings.py::ClientSettings -- sem isto, sobrepor
  // API_BASE_URL no ambiente funcionava pro cliente Python mas era
  // ignorado aqui do lado Ink.
  if (process.env.API_BASE_URL) {
    _apiBaseUrl = process.env.API_BASE_URL;
    return _apiBaseUrl;
  }
  const envPath = path.join(REPO_ROOT, ".env");
  const text = fs.readFileSync(envPath, "utf-8");
  const match = text.match(/^API_BASE_URL=(.*)$/m);
  if (!match) throw new Error(`API_BASE_URL não encontrado em ${envPath}`);
  _apiBaseUrl = match[1].trim();
  return _apiBaseUrl;
}

function tokenPath() {
  // Tem que ser o mesmo arquivo que token_store.py usa, não uma cópia --
  // senão `cronista login` (Python) e esta TUI (Node) teriam sessão
  // cada um por conta própria.
  const root = process.env.LOCALAPPDATA || os.homedir();
  return path.join(root, "cronista", "auth.json");
}

// Em cache depois da primeira leitura -- "Login" sempre sai do Ink e
// respawna um processo novo (ver index.js), então dentro da vida de um
// processo só este próprio saveTokens muda o arquivo; não há por que
// reabrir e reparsear auth.json a cada chamada autenticada (duas de uma
// vez, em paralelo, ao abrir uma reunião: getMeeting + getTranscript).
let _tokensCache;

function loadTokens() {
  if (_tokensCache !== undefined) return _tokensCache;
  const p = tokenPath();
  if (!fs.existsSync(p)) {
    _tokensCache = null;
    return _tokensCache;
  }
  const data = JSON.parse(fs.readFileSync(p, "utf-8"));
  _tokensCache = { accessToken: data.access_token, refreshToken: data.refresh_token };
  return _tokensCache;
}

function saveTokens(accessToken, refreshToken) {
  fs.writeFileSync(tokenPath(), JSON.stringify({ access_token: accessToken, refresh_token: refreshToken }));
  _tokensCache = { accessToken, refreshToken };
}

// fetch() nativo não tem timeout por padrão -- sem isso, uma porta que
// aceita conexão mas nunca responde trava a tela pra sempre.
const _TIMEOUT_MS = 10_000;

function _mensagemDeFalha(err) {
  if (err?.name === "TimeoutError") {
    return "Tempo esgotado conectando à API (10s). Verifique se o container está no ar e respondendo.";
  }
  return "Não foi possível conectar à API. Verifique se o container está no ar.";
}

async function request(method, apiPath, accessToken) {
  const base = readApiBaseUrl();
  let resp;
  try {
    resp = await fetch(`${base}${apiPath}`, {
      method,
      headers: accessToken ? { Authorization: `Bearer ${accessToken}` } : {},
      signal: AbortSignal.timeout(_TIMEOUT_MS),
    });
  } catch (err) {
    throw new ApiError(_mensagemDeFalha(err));
  }
  return resp;
}

async function refresh(refreshToken) {
  const base = readApiBaseUrl();
  let resp;
  try {
    resp = await fetch(`${base}/auth/refresh`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ refresh_token: refreshToken }),
      signal: AbortSignal.timeout(_TIMEOUT_MS),
    });
  } catch (err) {
    throw new ApiError(_mensagemDeFalha(err));
  }
  if (resp.status === 401) throw new ApiError("Sessão expirada. Faça login novamente.", 401);
  if (!resp.ok) throw new ApiError(`Falha ao renovar sessão (HTTP ${resp.status}).`, resp.status);
  const body = await resp.json();
  return body.access_token;
}

// 401 tenta renovar o token uma vez antes de desistir.
export async function authedGet(apiPath) {
  const tokens = loadTokens();
  if (!tokens) throw new ApiError("Não autenticado. Rode `cronista login`.", 401);

  let resp = await request("GET", apiPath, tokens.accessToken);
  if (resp.status === 401) {
    const newAccessToken = await refresh(tokens.refreshToken);
    saveTokens(newAccessToken, tokens.refreshToken);
    resp = await request("GET", apiPath, newAccessToken);
  }
  if (!resp.ok) throw new ApiError(`Erro na API (HTTP ${resp.status}).`, resp.status);
  return resp.json();
}

export function listMeetings() {
  return authedGet("/meetings");
}

export function getMeeting(meetingId) {
  return authedGet(`/meetings/${meetingId}`);
}

export function getTranscript(meetingId) {
  return authedGet(`/meetings/${meetingId}/transcript`);
}

export function search(q) {
  return authedGet(`/search?${new URLSearchParams({ q }).toString()}`);
}
